File: silver/4949/c++/main.cpp
#include <iostream>
#include <string>
#include <stack>

int main()
{
    std::string text;

    while (text != ".") {
        std::stack<char> array; 

        std::getline(std::cin, text);
        if (text == ".") {
            break;
        }
            
        for (char data : text) {
            if (array.size() == 0 && (data == ']' || data == ')')) {
                array.push('n');
                break;
            }

            if (!(data == '[' || data == ']' || data == '(' || data == ')')) {
                continue;
            }


            if (data == '[' || data == '(') {
                array.push(data);
            } else if (data == ']' && array.top() == '[') {
                array.pop();
            } else if (data == ')' && array.top() == '(') {
                array.pop();
            } else {
                array.push(data);
            }
        }

        std::cout << (array.size() == 0 ? "yes" : "no") << '\n';
    }

    return 0;
}

File: silver/4949/python/main.py

if (__name__ == "__main__"):

    while (1):
        stack = list()
        text = input()
        if (text == '.'):
            break
        
        for data in text:
             if ((not stack) and (data == ']' or data == ')')):
                stack = [1]
                break
             if (data in ['[', ']', '(', ')']):
                if (data in ['[', '(']):
                    stack.append(data)
                elif (data == ']' and stack[-1] == '['):
                    stack.pop(-1)
                elif (data == ')' and stack[-1] == '('):
                    stack.pop(-1)
                else:
                    stack.append(data)
        print("yes" if (not stack) else "no")

